### PR TITLE
retry and get workspace layout list functionality added

### PR DESCRIPTION
**Pull request by nerudadhich**

### CODE REVIEW

File: commands/cli/workspace.go [hunk 1]
```diff
@@ -77,6 +84,43 @@ func workspaceAll(_ *kingpin.ParseContext) error {
...
+	}
+
+	for _, p := range vars.Providers {
+		prettyPrint(p)
+	}
+
+	req := server.GetWorkspaceLayoutsRequest{Id: wid}
+
+	wL, err := client.GetWorkspaceLayouts(makeContext(nil, nil), &req)
+	if err != nil {
```
Comment (Talina06) [inline commands/cli/workspace.go]: err should be immutable. Rename to something else.

File: server/handler.go [hunk 2]
```diff
@@ -128,6 +123,26 @@ func (s *Server) getWorkspace(id string) (*Workspace, error) {
...
+func (s *Server) GetWorkspaceLayouts(ctx context.Context, in *GetWorkspaceLayoutsRequest) (*Layouts, error) {
+	keys, err := s.store.GetKeys(types.WORKSPACE+"/"+in.Id+"/"+types.LAYOUT+"/", "/")
+	if err != nil {
+		return nil, err
+	}
+
+	var layouts []*Layout
+	for _, k := range keys {
+		splits := strings.Split(k, "/")
+		if len(splits) != 5 {
```
Comment (Talina06) [inline server/handler.go]: The assumption of 5 is tricky. Can this be avoided by directly fetching layout IDs in someway?

File: commands/cli/workspace.go (hunk 1 above)
Comment (sagarrakshe) [reply commands/cli/workspace.go]: `err` is okay to be mutable.

File: commands/cli/layout.go [hunk 3]
```diff
@@ -241,6 +241,7 @@ func addLayoutCommands(app *kingpin.Application) {
 	dl := lCLI.Command("destroy", "Destroy layout").Action(clm.layoutDestroy)
 
 	al.Flag("dry", "Dry apply for in memory plan").BoolVar(&clm.dry)
+	al.Flag("retry", "Number of retry on layout apply").Default(defaultAttempts).Int64Var(&clm.retry)
```
Comment (mohandutt134) [inline commands/cli/layout.go]:
```suggestion
	al.Flag("retry", "Number of retries on layout apply, make it 0 for no retries, default is 3").Default(defaultAttempts).Int64Var(&clm.retry)
```

File: server/handler.go [hunk 4]
```diff
@@ -128,6 +123,26 @@ func (s *Server) getWorkspace(id string) (*Workspace, error) {
 	return &w, err
 }
 
+func (s *Server) GetWorkspaceLayouts(ctx context.Context, in *GetWorkspaceLayoutsRequest) (*Layouts, error) {
```
Comment (mohandutt134) [inline server/handler.go]: Need to rewrite this function to make the code simpler.

File: server/handler_test.go [hunk 5]
```diff
@@ -428,3 +475,149 @@ func TestServer_GetOutput(t *testing.T) {
...
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, layoutId, job.LayoutId)
+		assert.Equal(t, int32(JobState_PENDING), job.Status)
+		assert.Equal(t, int32(Operation_APPLY), job.Op)
+		assert.Equal(t, false, job.Dry)
+		assert.NotEmpty(t, job.LayoutVersion)
+		assert.Equal(t, int64(4), job.Retry)
+	})
```
Comment (mohandutt134) [inline server/handler_test.go]: Add unlock code here.

File: server/handler_test.go [hunk 6]
```diff
@@ -428,3 +475,149 @@ func TestServer_GetOutput(t *testing.T) {
...
+
+	lockKey := fmt.Sprintf("%v-%v", workspaceId, layoutId)
+
+	t.Run("Unlocking is Idempotent", func(t *testing.T) {
```
Comment (mohandutt134) [inline server/handler_test.go]: Not needed

**Pull request merged**